jolt-core/src/main/java/com/bazaarvoice/jolt/templatr/function/Math.java

```diff
@@ -18,151 +18,316 @@
 
 import com.bazaarvoice.jolt.common.Optional;
 
+import java.util.ArrayList;
+import java.util.List;
+
 @SuppressWarnings( "unused" )
 public class Math {
 
-    public static final class MaxOf implements Function {
-        @Override
-        public Optional<Object> apply( final Object... args ) {
-            Integer maxInt = Integer.MIN_VALUE;
-            Double maxDouble = -(Double.MIN_VALUE);
-            Long maxLong = Long.MIN_VALUE;
+    public static Optional<Number> MaxOf( Object... args ) {
+        if(args.length == 0) {
+            return Optional.empty();
+        }
 
-            if(args.length == 0) {
-                return Optional.empty();
+        Integer maxInt = Integer.MIN_VALUE;
+        Double maxDouble = -(Double.MAX_VALUE);
+        Long maxLong = Long.MIN_VALUE;
+
+        for(Object arg: args) {
+            if(arg instanceof Integer) {
+                maxInt = java.lang.Math.max( maxInt, (Integer) arg );
+            }
+            else if(arg instanceof Double) {
+                maxDouble = java.lang.Math.max( maxDouble, (Double) arg );
             }
+            else if(arg instanceof Long) {
+                maxLong = java.lang.Math.max(maxLong, (Long) arg);
+            }
+        }
+        if(maxInt == Integer.MIN_VALUE && maxDouble == ( - Double.MAX_VALUE ) && maxLong == Long.MIN_VALUE) {
+            return Optional.empty();
```

Comment:
BUG : if my input was Integer.MIN_VALUE then you would return Optional.empty(), stead of MIN_VALUE.

Need to add a boolean, "was anything found"? and only if that is false, return Optional.empty()